### PR TITLE
this moves aql::Query* out of the call parameter to any aql::Function

### PR DESCRIPTION
**Pull request by jsteemann**
it instead calls aql::Function will the much smaller ExpressionContext
interface, that only exposes a few functions that can refer to an
aql::Query* later. The interface can be specialized for several cases.
Right now the main implementation just uses aql::Query* under the hood,
but I plan to introduce another more lightweight implementation in
some other future use case.

This PR is just the preparatory work for that.

### CODE REVIEW

**Comment by jsteemann** [comment]
http://jenkins01.arangodb.biz:8080/job/arangodb-matrix-pr/1091/

**Comment by jsteemann** [comment]
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/1108/
(blue except unrelated failure)

File: arangod/Aql/QueryExpressionContext.h [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+  void registerWarning(int errorCode, char const* msg) override;
+  void registerError(int errorCode, char const* msg) override;
+  
+  icu::RegexMatcher* buildRegexMatcher(char const* ptr, size_t length, bool caseInsensitive) override;
+  icu::RegexMatcher* buildLikeMatcher(char const* ptr, size_t length, bool caseInsensitive) override;
+  icu::RegexMatcher* buildSplitMatcher(AqlValue splitExpression, transaction::Methods*, bool& isEmptyExpression) override;
+
+  bool killed() const override;
+  TRI_vocbase_t& vocbase() const override;
+  Query* query() const override;
```
Comment (graetzer) [inline arangod/Aql/QueryExpressionContext.h]: can you make these final ?

Comment (jsteemann) [reply arangod/Aql/QueryExpressionContext.h]: ok, did it.

**Pull request merged**